### PR TITLE
Use EMA instead of vector for e2e metric

### PR DESCRIPTION
**Pull request by lherman-cs**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by disa6302** [comment]
How does changing to sum and count help?

**Comment by lherman-cs** [comment]
@disa6302 It doesn't it's just an enhancement. But, the lock here https://github.com/aws-samples/amazon-kinesis-video-streams-demos/pull/122/files#diff-36c06b4e0a11b29790dda910ed5a0beb1cf833082548e8ebe1370d8be1c8e476R703, will help with the race condition.

File: canary/webrtc-c/src/CloudwatchMonitoring.cpp [hunk 1]
```diff
@@ -233,26 +233,28 @@ VOID CloudwatchMonitoring::pushInboundRtpStats(Canary::PIncomingRTPMetricsContex
...
+    DOUBLE latency = ctx.frameLatencyAvg / (DOUBLE) HUNDREDS_OF_NANOS_IN_A_MILLISECOND;
+
+    // TODO: due to https://github.com/aws-samples/amazon-kinesis-video-streams-demos/issues/96,
+    //       it's not clear why the emitted metric shows -nan. Since -nan is a string that's outputted
+    //       from Datum's value stringify implementation, we should try to get the original value by
+    //       printing it ourself.
+    //
+    //       If the issues doesn't exist anymore, please remove this as this is intended for debugging only.
+    //       The generic metric logging should be sufficient.
+    DLOGD("Current end-to-end frame latency: %4.2lf", latency);
```
Comment (disa6302) [inline canary/webrtc-c/src/CloudwatchMonitoring.cpp:248]: This is logged in the common functionality right? We will still be logging the aggregate here.

**Review by MushMal — APPROVED** [review]

File: canary/webrtc-c/src/Peer.cpp [hunk 2]
```diff
@@ -536,11 +536,16 @@ STATUS Peer::addTransceiver(RtcMediaStreamTrack& track)
         frameDataPtr += SIZEOF(UINT64);
         UINT32 receivedSize = getUnalignedInt32BigEndian((PINT32)(frameDataPtr));
 
-        pPeer->endToEndMetricsContext.frameLatency.push_back((DOUBLE)(GETTIME() - receivedTs) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+        pPeer->endToEndMetricsContext.frameLatencyAvg =
+            EMA_ACCUMULATOR_GET_NEXT(pPeer->endToEndMetricsContext.frameLatencyAvg, GETTIME() - receivedTs);
```
Comment (disa6302) [inline canary/webrtc-c/src/Peer.cpp:540]: Just for my understanding. What is the hypothesis for EMA fixing the e2e metric emitting problem?

**Review by hassanctech — APPROVED** [review]

File: canary/webrtc-c/src/CloudwatchMonitoring.cpp (hunk 1 above)
Comment (lherman-cs) [reply canary/webrtc-c/src/CloudwatchMonitoring.cpp:248]: It's as what the comment says, the common logging uses C++ string, which is entirely dependent on the implementor. What we want is the original value.

File: canary/webrtc-c/src/Peer.cpp (hunk 2 above)
Comment (lherman-cs) [reply canary/webrtc-c/src/Peer.cpp:540]: Again, it's not the EMA or how it's calculated, but there's a data race in collecting and publishing these metrics. Therefore, I added this lock, https://github.com/aws-samples/amazon-kinesis-video-streams-demos/pull/122/files#diff-36c06b4e0a11b29790dda910ed5a0beb1cf833082548e8ebe1370d8be1c8e476R702.

EMA is an enhancement from what we had, and should be preferred as it's standard in our SDKs to calculate average in a stream.

Comment (disa6302) [reply canary/webrtc-c/src/Peer.cpp:540]: Ok. Makes sense.

File: canary/webrtc-c/src/CloudwatchMonitoring.cpp (hunk 1 above)
Comment (disa6302) [reply canary/webrtc-c/src/CloudwatchMonitoring.cpp:248]: Ah! Ok. Now I get it.

**Review by disa6302 — APPROVED** [review]

**Pull request merged**